src/Google.Bigquery.V2/BigqueryClientImpl.DatasetCrud.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright 2016 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Apis.Bigquery.v2;
+using Google.Apis.Bigquery.v2.Data;
+using Google.Apis.Requests;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+
+namespace Google.Bigquery.V2
+{
+    public partial class BigqueryClientImpl
+    {
+        private static readonly PageStreamer<DatasetList.DatasetsData, DatasetsResource.ListRequest, DatasetList, string> s_datasetsPageStreamer =
+            new PageStreamer<DatasetList.DatasetsData, DatasetsResource.ListRequest, DatasetList, string>(
+            (request, token) => request.PageToken = token,
+            response => response.NextPageToken,
+            response => response.Datasets);
+
+        /// <inheritdoc />
+        public override BigqueryDataset GetDataset(DatasetReference datasetReference)
+        {
+            Preconditions.CheckNotNull(datasetReference, nameof(datasetReference));
+            var resource = Service.Datasets.Get(datasetReference.ProjectId, datasetReference.DatasetId).Execute();
+            return new BigqueryDataset(this, resource);
+        }
+
+        /// <inheritdoc />
+        public override IEnumerable<BigqueryDataset> ListDatasets(string projectId, ListDatasetsOptions options = null)
```

Comment:
Currently only implementing this in the same way as for StorageClient. We can then work out what to do with "all REST-based wrappers" in one go.